Salesforce/client/token_refresher.py

```diff
@@ -0,0 +1,160 @@
+    _locks: dict[str, Lock] = {}
+    _session: ClientSession | None = None
+
+    def __init__(self, client_id: str, client_secret: str, auth_url: str, token_ttl: int = 3600):
+        """
+        Initialize SalesforceTokenRefresher.
+
+        Each set of scopes will have its own access token with refreshing logic.
+
+        Args:
+            client_id: str
+            client_secret: str
+            auth_url: str
+            token_ttl: int
+        """
+        self.client_id = client_id
+        self.client_secret = client_secret
+        self.auth_url = auth_url
+        self.token_ttl = token_ttl
+
+        self._token: SalesforceToken | None = None
+        self._token_refresh_task: Optional[Task[None]] = None
+
+    @classmethod
+    def session(cls) -> ClientSession:
+        """
+        Initialize client session.
+
+        Singleton client session to work with token refresh logic.
+
+        Returns:
+            ClientSession:
+        """
+        if not cls._session:
+            cls._session = ClientSession()
+
+        return cls._session
+
+    @classmethod
+    async def instance(
+        cls,
+        client_id: str,
+        client_secret: str,
+        auth_url: str,
+    ) -> "SalesforceTokenRefresher":
+        """
+        Get singleton SalesforceTokenRefresher instance for specified set of scopes.
+
+        Args:
+            client_id: str
+            client_secret: str
+            auth_url: str
+
+        Returns:
+            SalesforceTokenRefresher:
+        """
+        refresher_unique_key = str(frozenset({client_id, client_secret}))
+        if not cls._locks.get(refresher_unique_key):
+            cls._locks[refresher_unique_key] = asyncio.Lock()
+
+        if not cls._instances.get(refresher_unique_key):
+            async with cls._locks[refresher_unique_key]:
+                if not cls._instances.get(refresher_unique_key):
+                    cls._instances[refresher_unique_key] = SalesforceTokenRefresher(
+                        client_id,
+                        client_secret,
+                        auth_url,
+                    )
+
+        return cls._instances[refresher_unique_key]
+
+    async def refresh_token(self) -> None:
+        """
+        Refresh token based on class configuration.
+
+        Also triggers token refresh task.
+        """
+        params = {
+            "grant_type": "client_credentials",
+        }
+
+        url = URL("{0}/services/oauth2/token".format(self.auth_url)).with_query(urlencode(params, encoding="utf-8"))
+
+        async with self.session().post(url, auth=BasicAuth(self.client_id, self.client_secret), json={}) as response:
+            response_data = await response.json()
+
+            self._token = SalesforceToken(
+                token=HttpToken(**response_data),
+                created_at=time.time(),
+                ttl=self.token_ttl,
+            )
+
+            await self._schedule_token_refresh(self._token.ttl)
```

Comment:
How long last Salesforce's tokens?
We usually refresh the token 300 seconds (5 minutes) before it expires (e.g.: [CrowdStrike Auth](https://github.com/SEKOIA-IO/automation-library/blob/main/CrowdStrikeFalcon/crowdstrike_falcon/client/auth.py#L46)) to avoid negative responses due to authorization expiration.
```suggestion
            await self._schedule_token_refresh(self._token.ttl)
```

Reply:
Salesforce tokens have lifetime period of 2 hours. 
The problem is that they do not have any field in response that will identify expiration time, something like `expire_in`, `lifetime` or something similar.
Also there is an option to parse this token as JWT, it should have all info inside but I'm not sure if it is good idea
I'll setup refresh to 5 mins, but yes

Reply:
Yes. It is working solution. I've changed it to 5 mins, but token by itself is valid for 2 hours

Reply:
3600 was great. Just a mismatch between the comment and the implementation.